File: wisefy/src/androidTest/java/com/isupatches/wisefy/ConnectToNetworkTests.java
package com.isupatches.wisefy;


import android.net.wifi.WifiConfiguration;
import com.isupatches.wisefy.callbacks.ConnectToNetworkCallbacks;
import com.isupatches.wisefy.constants.WiseFyCodes;
import org.junit.Test;
import static com.isupatches.wisefy.base.TestUtils.TEST_SSID;
import static com.isupatches.wisefy.base.TestUtils.TEST_TIMEOUT;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;
import static org.mockito.ArgumentMatchers.anyBoolean;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.after;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.timeout;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;


public class ConnectToNetworkTests extends BaseAndroidJUnit4TestClass {

    @Test
    public void sync_failure_nullSSIDParam() {
        assertEquals(false, mWiseFy.connectToNetwork(null, TEST_TIMEOUT));
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).enableNetwork(anyInt(), anyBoolean());
    }

    @Test
    public void sync_failure_missingPrerequisites() {
        missingPrerequisite();
        assertEquals(false, mWiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT));
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).enableNetwork(anyInt(), anyBoolean());
    }

    @Test
    public void sync_failure_noSavedNetwork() {
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(null);

        assertEquals(false, mWiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT));
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).enableNetwork(anyInt(), anyBoolean());
    }

    @Test
    public void sync_failure() {
        WifiConfiguration wiFiConfiguration = new WifiConfiguration();
        wiFiConfiguration.SSID = TEST_SSID;
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(wiFiConfiguration);

        WiseFyConnection mockWiseFyConnection = mock(WiseFyConnection.class);
        mWiseFy.mWiseFyConnection = mockWiseFyConnection;
        when(mockWiseFyConnection.waitToConnectToSSID(anyString(), anyInt())).thenReturn(false);

        assertEquals(false, mWiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT));
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).enableNetwork(anyInt(), anyBoolean());
    }

    @Test
    public void sync_success() {
        WifiConfiguration wiFiConfiguration = new WifiConfiguration();
        wiFiConfiguration.SSID = TEST_SSID;
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(wiFiConfiguration);

        WiseFyConnection mockWiseFyConnection = mock(WiseFyConnection.class);
        mWiseFy.mWiseFyConnection = mockWiseFyConnection;
        when(mockWiseFyConnection.waitToConnectToSSID(anyString(), anyInt())).thenReturn(true);

        assertEquals(true, mWiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT));
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).enableNetwork(anyInt(), anyBoolean());
    }

    @Test
    public void async_failure_nullSSIDParam() {
        ConnectToNetworkCallbacks mockCallbacks = mock(ConnectToNetworkCallbacks.class);
        mWiseFy.connectToNetwork(null, TEST_TIMEOUT, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).connectToNetworkWiseFyFailure(WiseFyCodes.MISSING_PARAMETER);
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).enableNetwork(anyInt(), anyBoolean());
    }

    @Test
    public void async_failure_nullSSIDParam_nullCallback() {
        try {
            mWiseFy.connectToNetwork(null, TEST_TIMEOUT, null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).enableNetwork(anyInt(), anyBoolean());
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_failure_missingPrerequisites() {
        missingPrerequisite();
        ConnectToNetworkCallbacks mockCallbacks = mock(ConnectToNetworkCallbacks.class);
        mWiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).connectToNetworkWiseFyFailure(WiseFyCodes.MISSING_PREREQUISITE);
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).enableNetwork(anyInt(), anyBoolean());
    }

    @Test
    public void async_failure_missingPrerequisites_nullCallback() {
        missingPrerequisite();
        try {
            mWiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT, null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).enableNetwork(anyInt(), anyBoolean());
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_failure_noSavedNetwork() {
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(null);

        ConnectToNetworkCallbacks mockCallbacks = mock(ConnectToNetworkCallbacks.class);
        mWiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).networkNotFoundToConnectTo();
        verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).enableNetwork(anyInt(), anyBoolean());
    }

    @Test
    public void async_failure_noSavedNetwork_nullCallback() {
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(null);

        try {
            mWiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT, null);
            verify(mMockWiFiManager, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).enableNetwork(anyInt(), anyBoolean());
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_failure() {
        WifiConfiguration wiFiConfiguration = new WifiConfiguration();
        wiFiConfiguration.SSID = TEST_SSID;
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(wiFiConfiguration);

        WiseFyConnection mockWiseFyConnection = mock(WiseFyConnection.class);
        mWiseFy.mWiseFyConnection = mockWiseFyConnection;
        when(mockWiseFyConnection.waitToConnectToSSID(anyString(), anyInt())).thenReturn(false);

        ConnectToNetworkCallbacks mockCallbacks = mock(ConnectToNetworkCallbacks.class);
        mWiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).failureConnectingToNetwork();
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).enableNetwork(anyInt(), anyBoolean());
    }

    @Test
    public void async_failure_nullCallback() {
        WifiConfiguration wiFiConfiguration = new WifiConfiguration();
        wiFiConfiguration.SSID = TEST_SSID;
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(wiFiConfiguration);

        WiseFyConnection mockWiseFyConnection = mock(WiseFyConnection.class);
        mWiseFy.mWiseFyConnection = mockWiseFyConnection;
        when(mockWiseFyConnection.waitToConnectToSSID(anyString(), anyInt())).thenReturn(false);

        try {
            mWiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT, null);
            verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).enableNetwork(anyInt(), anyBoolean());
        } catch (NullPointerException npe) {
            fail();
        }
    }

    @Test
    public void async_success() {
        WifiConfiguration wiFiConfiguration = new WifiConfiguration();
        wiFiConfiguration.SSID = TEST_SSID;
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(wiFiConfiguration);

        WiseFyConnection mockWiseFyConnection = mock(WiseFyConnection.class);
        mWiseFy.mWiseFyConnection = mockWiseFyConnection;
        when(mockWiseFyConnection.waitToConnectToSSID(anyString(), anyInt())).thenReturn(true);

        ConnectToNetworkCallbacks mockCallbacks = mock(ConnectToNetworkCallbacks.class);
        mWiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT, mockCallbacks);
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).connectedToNetwork();
        verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).enableNetwork(anyInt(), anyBoolean());
    }

    @Test
    public void async_success_nullCallback() {
        WifiConfiguration wiFiConfiguration = new WifiConfiguration();
        wiFiConfiguration.SSID = TEST_SSID;
        WiseFySearch mockWiseFySearch = mock(WiseFySearch.class);
        mWiseFy.mWiseFySearch = mockWiseFySearch;
        when(mockWiseFySearch.findSavedNetworkByRegex(anyString())).thenReturn(wiFiConfiguration);

        WiseFyConnection mockWiseFyConnection = mock(WiseFyConnection.class);
        mWiseFy.mWiseFyConnection = mockWiseFyConnection;
        when(mockWiseFyConnection.waitToConnectToSSID(anyString(), anyInt())).thenReturn(true);

        try {
            mWiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT, null);
            verify(mMockWiFiManager, timeout(VERIFICATION_SUCCESS_TIMEOUT)).enableNetwork(anyInt(), anyBoolean());
        } catch (NullPointerException npe) {
            fail();
        }
    }
}

File: wisefy/src/androidTest/java/com/isupatches/wisefy/main/ConnectToNetworkTests.kt
package com.isupatches.wisefy.main

import com.isupatches.wisefy.BaseAndroidJUnit4TestClass
import com.isupatches.wisefy.TEST_SSID
import com.isupatches.wisefy.TEST_TIMEOUT
import com.isupatches.wisefy.VERIFICATION_SUCCESS_TIMEOUT
import com.isupatches.wisefy.callbacks.ConnectToNetworkCallbacks
import com.isupatches.wisefy.constants.MISSING_PARAMETER

import org.junit.Assert.assertEquals
import org.junit.Test

import org.mockito.Mockito.mock
import org.mockito.Mockito.timeout
import org.mockito.Mockito.verify

/**
 * Tests the ability to connect to a network.
 *
 * @author Patches
 */
internal class ConnectToNetworkTests : BaseAndroidJUnit4TestClass() {

    @Test fun sync_failure_prechecks() {
        mockWiseFyPrechecksUtil.connectToNetwork_failure()
        assertEquals(false, wiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT))
        verificationUtil.didNotTryToConnectToNetwork()
    }

    @Test fun sync_failure_noSavedNetwork() {
        mockWiseFyPrechecksUtil.connectToNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_null()
        assertEquals(false, wiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT))
        verificationUtil.didNotTryToConnectToNetwork()
    }

    @Test fun sync_failure() {
        mockWiseFyPrechecksUtil.connectToNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        mockWiseFyConnectionUtil.waitToConnectToSSID(false)
        assertEquals(false, wiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT))
        verificationUtil.triedToConnectToNetwork()
    }

    @Test fun sync_success() {
        mockWiseFyPrechecksUtil.connectToNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        mockWiseFyConnectionUtil.waitToConnectToSSID(true)
        assertEquals(true, wiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT))
        verificationUtil.triedToConnectToNetwork()
    }

    @Test fun async_failure_prechecks() {
        mockWiseFyPrechecksUtil.connectToNetwork_failure()
        val mockCallbacks = mock(ConnectToNetworkCallbacks::class.java)
        wiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wisefyFailure(MISSING_PARAMETER)
        verificationUtil.didNotTryToConnectToNetwork()
    }

    @Test fun async_failure_prechecks_nullCallback() {
        mockWiseFyPrechecksUtil.connectToNetwork_success()
        mockWiseFyPrechecksUtil.connectToNetwork_failure()
        nullCallbackUtil.callConnectToNetwork(TEST_SSID)
        verificationUtil.didNotTryToConnectToNetwork()
    }

    @Test fun async_failure_noSavedNetwork() {
        mockWiseFyPrechecksUtil.connectToNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_null()
        val mockCallbacks = mock(ConnectToNetworkCallbacks::class.java)
        wiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).networkNotFoundToConnectTo()
        verificationUtil.didNotTryToConnectToNetwork()
    }

    @Test fun async_failure_noSavedNetwork_nullCallback() {
        mockWiseFyPrechecksUtil.connectToNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_null()
        nullCallbackUtil.callConnectToNetwork(TEST_SSID)
        verificationUtil.didNotTryToConnectToNetwork()
    }

    @Test fun async_failure() {
        mockWiseFyPrechecksUtil.connectToNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        mockWiseFyConnectionUtil.waitToConnectToSSID(false)
        val mockCallbacks = mock(ConnectToNetworkCallbacks::class.java)
        wiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).failureConnectingToNetwork()
        verificationUtil.triedToConnectToNetwork()
    }

    @Test fun async_failure_nullCallback() {
        mockWiseFyPrechecksUtil.connectToNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        mockWiseFyConnectionUtil.waitToConnectToSSID(false)
        nullCallbackUtil.callConnectToNetwork(TEST_SSID)
        verificationUtil.triedToConnectToNetwork()
    }

    @Test fun async_success() {
        mockWiseFyPrechecksUtil.connectToNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        mockWiseFyConnectionUtil.waitToConnectToSSID(true)
        val mockCallbacks = mock(ConnectToNetworkCallbacks::class.java)
        wiseFy.connectToNetwork(TEST_SSID, TEST_TIMEOUT, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).connectedToNetwork()
        verificationUtil.triedToConnectToNetwork()
    }

    @Test fun async_success_nullCallback() {
        mockWiseFyPrechecksUtil.connectToNetwork_success()
        mockWiseFySearchUtil.findSavedNetworkByRegex_success()
        mockWiseFyConnectionUtil.waitToConnectToSSID(true)
        nullCallbackUtil.callConnectToNetwork(TEST_SSID)
        verificationUtil.triedToConnectToNetwork()
    }
}
